### DISCUSSION TITLE
Enable Disabling Preflight Based on Path

### DISCUSSION THREAD

**Discussion by Blushyes** [tailwindlabs/tailwindcss#12624] [Ideas]
I've been working on a `Chrome extension` and encountered a challenging issue with the `preflight` feature of Tailwind CSS, which tends to disrupt the styling of the original web page. Currently, there doesn't seem to be built-in support for selectively disabling the `preflight` feature at the `content-script` level while keeping it enabled elsewhere, such as in `popups`.

As a workaround, I have stopped including `@tailwind base` in the `content script`, but this approach occasionally leads to incorrect styles from Tailwind CSS. I am wondering if there is an existing solution to achieve such isolation, or if this could be a feature worth considering for addition to Tailwind CSS to better support Chrome extension development.

It likes:
<img width="756" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/75828899/777a5f98-a56c-4ef6-ac05-dc459a795286">

But this is what it originally looked like:
<img width="857" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/75828899/83677f1a-390a-4b14-bdb6-5113dcaed0bb">

**Comment by wongjn** [comment]
For what it's worth, you'd want to [disable the preflight plugin](https://tailwindcss.com/docs/preflight#disabling-preflight), instead of removing `@tailwind base`.

**Reply by Blushyes** [reply]
This will result in preflight being globally turned off, which might cause issues with the normal Tailwind styles in my other environments such as popup. However, I just want to disable preflight for the content-script environment only. The directory structure looks roughly like this:

- src
  - content-script (who will cause the source webpage to become disorganized)
  - popup 
  
I only want to disable preflight for the files within content-script, not for my entire project.

**Reply by wongjn** [reply]
Thank, I understood your requirements the first time around! My point was that you'd want to [disable the preflight plugin](https://tailwindcss.com/docs/preflight#disabling-preflight), instead of removing `@tailwind base` – not that this would fulfil your requirements.

As a work-around, you could [disable the preflight plugin](https://tailwindcss.com/docs/preflight#disabling-preflight) and [reimplement the core code](https://github.com/tailwindlabs/tailwindcss/blob/8350cffdad34efb6625e0f2b0ff738edaf0d1cce/src/corePlugins.js#L507-L518) and scope it yourself within [your own Tailwind plugin](https://tailwindcss.com/docs/plugins#adding-base-styles).

**Reply by Blushyes** [reply]
👌🏻, thanks, I will try it.

**Reply by Blushyes** [reply]
I just gave it a try to write a Tailwind Plugin like this, but unfortunately, it seems that Tailwind does not provide the capability to determine the path of the file it is currently processing. Thus, the only option appears to be attempting to write a plugin for the build tool, which is obviously quite complex. So, it seems that the issue at hand is whether Tailwind needs to support plugins being aware of which files they are processing.

```js
const fs = require('fs')
const path = require('path')
const postcss = require('postcss')
const tailwindVersion = require('tailwindcss/package.json').version

module.exports = function customPreflightPlugin() {
  return function ({ addBase }) {
    const shouldApplyPreflight = (context) => {
      // The logic can be based on the file path provided in the context
      // For example, do not apply preflight for files under `src/content-script` directory
      // However, the file path information in this context is not always available, it depends on how you run Tailwind
      if (!context) return true
      return !context.includes('./content-script')
    }

    const tailwindCssPath = path.dirname(
      require.resolve('tailwindcss/package.json')
    )
    const preflightPath = path.join(tailwindCssPath, 'src/css/preflight.css')
    let preflightStyles = postcss.parse(fs.readFileSync(preflightPath, 'utf8'))

    const addCustomPreflight = (context) => {
      if (!shouldApplyPreflight(context)) {
        return
      }
      addBase([
        postcss.comment({
          text: `! tailwindcss v${tailwindVersion} | MIT License | https://tailwindcss.com`,
        }),
        ...preflightStyles.nodes,
      ])
    }

    addCustomPreflight(['...', ...]) // Here we should pass the file path parameter
  }
}
```

**Reply by wongjn** [reply]
You could use the [`@config`](https://tailwindcss.com/docs/functions-and-directives#config) directive to set a different Tailwind configuration file for the CSS file you'd want your modified preflight in. Then, have this Tailwind configuration be nearly the exact same as your main Tailwind configuration file (or have your main Tailwind configuration as [a preset](https://tailwindcss.com/docs/presets)) and add your modified preflight plugin to only that Tailwind configuration.

**Reply by Blushyes** [reply]
Oh my dude, the problem that's been bugging me all day is finally solved! I'm super grateful to you! Looks like diving into the official docs really pays off when it comes to these nitty-gritty details! 📚✨
